Implementing Flask API using Linked list Data Structure (#3880)
section-engineering-education/engineering-education: content/articles/implementing-flask-api-using-linked-list/index.md

### Implementing Flask API using Linked list Data Structure

> ![hero_image](/engineering-education/implementing-flask-api-using-linked-list/hero.jpg)

Reviewer: Hello, the hero image is not supposed to be in the article itself. It is only required in the article folder. Please delete this part.😊

5. [Testing the API using Postman](#testing-the-api-using-postman)
6. [Conclusion](#conclusion)

### Prerequisites
Before we start building the application, you should have the latest version of [Python](https://www.python.org/downloads/) installed or at least version `Python 3.6+`.

To help install the necessary packages and dependencies you should have `pip` installed. If not check out the [link](https://pip.pypa.io/en/stable/installation/) to have it in your machine depending on your operating system.

We'll be using the DB Browser for SQLite to visually check the database and its tables. You can download it from [here](https://sqlitebrowser.org/).

For testing APIs you will need [Postman](https://www.postman.com/downloads/). Depending on your preference you can choose a text editor of your choice.

There are a number such as [Visual Studio Code](https://code.visualstudio.com/download), [Atom](https://atom.io/), [Sublime](https://www.sublimetext.com/3), [Vim](https://www.vim.org/download.php) among others.

For this tutorial, I'll be using Ubuntu 20.04 and Vim text editor.  

### Introduction to Flask

> Flask is a micro-framework of Python that is used to create custom web applications. It is lightweight and does not come with any bundled packages and dependencies enabling developers to be on the wheel while building any application on it.

Reviewer: Use the correct punctuation. The second sentence seems to be too long. Break it down using commas or make it to two sentences.

The `self.data` and `self.next_node` either default to None or equals the value pass to data or next_node during the time of instantiation. The `__init__()` method is a constructor and it allows the class to initialize its attributes.

The `LinkedList()` class constructor takes `head` and `last_node` as its attribute and assign them to None. We then create four methods `to_list()`, `insert_beginning`, `insert_at_end` ,and `print_ll` which we'll implement in a moment.

With that said, let's visualize how a linked list works:

![linked_list](/engineering-education/implementing-flask-api-using-linked-list/linked-list.jpg)

The rectangles in the image above represent nodes in a linked list and each node has two separate compartments.

The left compartment represents the data and has the string value Data in it. The right compartment represents the pointer that points to the next node.

If we look at our Node class we see it has the data and next_node pointer as well. The `LinkedList()` wrapper class only helps us keep track of the head of our linked list.

> From the image above the first node in the linked list represents the head. So for instance, if we want to add to the head of our linked list as we'll see in a bit it will be easier to do so.

Reviewer: Add a comma to the first sentence.

The `toList()` takes a linked list object and converts it into a list. The implementation is shown below:

```python
def to_list(self):
  l = []
  if self.head is None:
    return l
  node = self.head
  while node:
    l.append(node.data)
    node = node.next_node
  return l
```

First, we create an empty list. Then proceed to check if the head of the linked list is `None`.

> If so, we return the empty list otherwise it's appended onto the list and the node is assigned to the node's next node. The list is then returned.

Reviewer: ```suggestion
If so, we return the empty list. Otherwise it is appended onto the list and the node is assigned to the node's next node. The list is then returned.
```

The `insert_beginning()` implementation is shown below:

```python
def insert_beginning(self, data):
new_node = Node(data, self.head)
self.head = new_node
```

> It takes `self` and `data` as arguments. This means that if data is added at the beginning of our linked list, it assumes being the head and the current head becomes the next_node pointer.

Reviewer: What takes `self` and `data` as arguments?

We then assign `new_node` as the head using `self` to instantiate it.
We will write a function to print the testcases later, for now, let's implement the `insert_at_end()`.

```python
def insert_at_end(self, data):
if self.head is None:
return self.insert_beginning(data)

node = self.head
while node.next_node:
node = node.next_node
node.next_node = Node(data, None)
```

> The above code checks if the current head is none. If so, it calls the `insert_beginning()` to insert data and return it. The head value is assigned to a `node` variable and it performs a while loop.

Reviewer: ```suggestion
The above code checks if the current head is none. If so, it then calls the `insert_beginning()` to insert data and return it. The head value is assigned to a `node` variable and it performs a while loop.
```

> If the node's next node is true, it assign it to the node. otherwise, it adds data to the node using the `Node` class.

Reviewer: ```suggestion
If the node's next node is true, it assigns it to the node. Otherwise, it adds data to the node using the `Node` class.
```

Let's now write a function to test the linked list functions:

```python
def print_ll(self):
ll_string = ''
node = self.head
if node is None:
print(None)
while node:
ll_string += f'{str(node.data)} ->'
node = node.next_node
ll_string += ' None'
print(ll_string)
```
From the above snippet:

> We start by creating an empty string called `ll_string` then assign the head to a variable  `node`. If the node is none; empty, the console prints `None`.

Reviewer: ```suggestion
We started by creating an empty string called `ll_string` then assigned the head to a variable  `node`. If the node is none; empty, the console prints `None`.
```

Now let's initialize the application by assigning the Flask instance to the `app` variable. This creates a flask object that implements the WSGI application and acts as the central registry for the application packages and modules.

Setting the configuration dictionary allow specification of the database path to our database using the `SQLALCHEMY_DATABASE_URI`. The `SQLALCHEMY_TRACK_MODIFICATIONS` is set to True as Flask-SQLAlchemy tracks the modification of objects and emits signals.

The configuration of the database takes a `listens_for` decorator from the event API which accepts the `Engine` and `"connect"` as arguments. The `"connect"` identifies the event to be intercepted, and a user-defined listening function.

The `_set_sqlite_pragma()` creates a connection based on our `sqlite3` database, enable the foreign key constraints, and closes the connection. For more on event registration check the [event documentation](https://docs.sqlalchemy.org/en/14/core/event.html).

The `db` variable creates the instance of the database.

Our application will have one model called `User` which we'll add its attributes in a few.

> We define five functions that will handle various logic and takes different parameters. The `create_user()` function takes the route decorator and pass two arguments the `/user` and `POST` HTTP request.

Reviewer: ```suggestion
We define five functions that will handle various logic and takes different parameters. The `create_user()` function takes the route decorator and passes two arguments the `/user` and `POST` HTTP request.
```

The `get_user()` fetches a single user of any given `id`. The `delete_user()` function deletes a specified user with given `id` and uses the `DELETE` HTTP request.

The file ends by setting the debug mode to `True`. This provides a useful debugger tool to track any errors in the application.

#### Models
The `User` model has various attributes as shown below:

```python
class User(db.Model):
  __tablename__ = 'users'
  id = db.Column(db.Integer, primary_key=True)
  name = db.Column(db.String(50))
  email = db.Column(db.String(200))
  address = db.Column(db.String(50))
  phone = db.Column(db.String(50))
```

The `__tablename__` overrides and sets the table name of our model. The `Column` keyword defines a column in the database table. The `id` attribute takes the type `Integer` since a primary key cannot be of any other type.

It is marked True as the `primary_key`. The `name`, `email`, `address`, and `phone` take the `String` type. The `String()` type takes the size parameter of maximum length.

> We can now generate our database based on the model we created by running our file through the python shell. Let's save the file and head to the terminal and execute the commands below:

Reviewer: ```suggestion
We can now generate our database, based on the model we created by running our file through the python shell. Let's save the file and head to the terminal and execute the commands below:
```

```bash
python
from server import db
db.create_all()
exit()
```
By executing the word `python` in the terminal it enters into an interactive shell and then executes the preceding commands.

> If we now `ls` to list the contents of the `FlaskAPI` directory we're able to see a `sqlitedb.file` file and that's our database.

Reviewer: Kindly explain this sentence, it does not make sense.

          name = data['name'],
          email = data['email'],
          address = data['address'],
          phone = data['phone']
  )
  db.session.add(new_user)
  db.session.commit()
  return jsonify({"message": "User created"}), 200
```

We had earlier covered what the decorator does. The `data` will hold requests being parsed in JSON format using the `get_json()` Flask API function.

The `new_user` variable is the instance of the User model and its assigned properties to the data in dictionary format. We then add the instance to the database and commit it.

If the user is created successfully a JSON formatted response will be displayed with the status `200`. We will then import the `jsonify` from the Flask's `flask.json` module.

It serializes data to JavaScript Object Notation(JSON) format.  We'll test this in a bit using `Postman` in the next chapter. For now let's create the routes.

In the `get_all_users_descending()`, `get_all_users_ascending()`, and `get_user()` we'll be using the linked list file that we created by importing the `LinkedList` class. So, in our `server.py` file import the `LinkedList` using:

```python
from linked_list import LinkedList
```

> Let's implement the endpoints.

Reviewer: ```suggestion
Let's implement the endpoints:
```

### Testing the API using Postman
To begin the testing the endpoints, fire up `Postman` and `server.py` file by the `python server.py` command.

Next, add a request called `create_user` and a `POST` request to it. To begin testing the `create_user()`, copy the running server and paste it to Postman and add the `/user` to appear like `http://127.0.0.1:5000/user`.

We then choose the `body` tab and enable `raw`. This allow us to select the `JSON` format for writing our data.

We write our request payload in the canvas space using the User model attributes, `name`, `email`, `address`, and `phone`. The image below shows a snap of that endpoint:

![create_user](/engineering-education/implementing-flask-api-using-linked-list/create-user.jpg)

We create at least five users to enable us to test other endpoints.
In order to avoid running into `sqlite3.OperationalError` close the DB Browser for SQLite when sending requests.

Performing the payload request `get_all_users_descending()`, `get_all_users_descending()`, and `get_user()` is easier as we just only specify the routes and select the `GET` request and send. This performs the logic implemented in each function.

> The `get_all_users_descending` fetch the users in descending order and `get_all_users_ascending` perform the same in ascending order.

Reviewer: ```suggestion
The `get_all_users_descending` fetches the users in descending order and `get_all_users_ascending` perform the same in ascending order.
```

The `get_user()` returns the payload of that particular `id` specified. The `delete_user()` deletes a user with the specified ID passed.

For example if you want to delete a user with the ID of 3, you pass `http://127.0.0.1:5000/user/3` with the `DELETE` request and send. If you check it in the DB Browser it won't be available.

### Conclusion

> To recap on what this article has covered, we first introduce what Flask is and a few reasons why it is most preferred.We then got a better understanding of what linked lists are and implemented some of its use-cases.

Reviewer: ```suggestion
To recap on what this article has covered, we first introduced what Flask is and a few reasons why it is most preferred. We then got a better understanding of what linked lists are and implemented some of its use-cases.
```

---
layout: engineering-education
status: publish
published: true
url: /implementing-flask-api-using-linked-list-data-structure/
title: Implementing Flask API Using Linked List Data Structure
description: In this tutorial, we will discuss about Flask and linked lists. We will also learn how we can use linked list class to implement the core API endpoints.
author: oruko-pius
date: 2021-03-03T00:00:00-13:00
topics: []
excerpt_separator: <!--more-->
images:

  - url: /engineering-education/implementing-flask-api-using-linked-list-data-structure/hero.jpg
    alt: Implementing Flask API Using Linked List Hero Image
---

### Introduction
Understanding data structures is an important aspect when developing applications. As real-world applications consider space and time complexities when designing applications to save on the memory and time taken to run specific programs.
<!--more-->

> Despite the useful trade-offs, they offer to the pool of developers, most rarely use them. These make their programs and software not being to the quality and standards expected and can make organizations incur huge losses in terms of budgeting for storage and consumer loading time of these applications.

Reviewer: hello @OkothPius, could you kindly look at this paragraph? The second sentence is too long and also is hard to understand. Correct the sentence and punctuate it.
Then, try to explain the concept in simple terms for easy understanding.
